src/main/java/Buu/Ui.java

```diff
@@ -0,0 +1,150 @@
+    }
+    /**
+     * Displays a confirmation message indicating that the task has been marked as done.
+     *
+     * @param task The task that has been marked as completed.
+     */
+    public void showTaskMarkedDone(Task task) {
+        outputBuffer.add("Nice! I've marked this task as done:");
+        outputBuffer.add("  " + task);
+    }
+    /**
+     * Displays a confirmation message indicating that the task has been unmarked as not done.
+     *
+     * @param task The task that has been unmarked as not completed.
+     */
+    public void showTaskUnmarked(Task task) {
+        outputBuffer.add("OK, I've marked this task as not done yet:");
+        outputBuffer.add("  " + task);
+    }
+
+    /**
+     * Displays a message to the user when a task is removed from the list.
+     *
+     * @param task      The task that was removed.
+     * @param taskCount The total number of tasks remaining in the list after removing the task.
+     */
+    public void showTaskRemoved(Task task, int taskCount) {
+        outputBuffer.add("Noted. I've removed this task:\n  "
+                + task + "\nNow you have " + taskCount + " tasks in the list.");
+    }
+
+    /**
+     * Displays the current list of tasks to the user.
+     *
+     * @param tasks The list of tasks to be displayed.
+     */
+    public void showTaskList(ArrayList<Task> tasks) {
+        if (tasks.isEmpty()) {
+            outputBuffer.add("No tasks to show.");
+        } else {
+            StringBuilder taskListMessage = new StringBuilder("Here are the tasks in your list:\n");
+            for (int i = 0; i < tasks.size(); i++) {
+                taskListMessage.append((i + 1)).append(".").append(tasks.get(i)).append("\n");
+            }
+            outputBuffer.add(taskListMessage.toString());
+        }
+    }
+
+    /**
+     * Displays the list of tasks that match the search keyword.
+     *
+     * @param tasks The list of tasks that match the search keyword.
+     */
+    public void showMatchingTasks(List<Task> tasks) {
+        if (tasks.isEmpty()) {
+            outputBuffer.add("No matching tasks found.");
+        } else {
+            StringBuilder matchingTasksMessage = new StringBuilder("Here are the matching tasks in your list:\n");
+            for (int i = 0; i < tasks.size(); i++) {
+                matchingTasksMessage.append((i + 1)).append(".").append(tasks.get(i)).append("\n");
+            }
+            outputBuffer.add(matchingTasksMessage.toString());
+        }
+    }
+
+    /**
+     * Clears the output buffer.
+     */
+    public void clearOutputBuffer() {
+        outputBuffer.clear();
+    }
+
+    /**
+     * Retrieves all the messages stored in the output buffer.
+     *
+     * @return The list of buffered messages.
+     */
+    public List<String> getOutputBuffer() {
+        return new ArrayList<>(outputBuffer);
+    }
+
+    public void addToOutputBuffer(String message) {
+        outputBuffer.add(message);
+    }
+}
```

Comment:
Very good application of OOP principles, with separation into modular components. The methods are very readable, and intuitive. It will be easy to extend this class further for more features involving UI.